# **VERİ DÜZENLEME ARACI (DPLYR) TEKRAR (FARKLI KAYNAK)**

Veri düzenlemede en sık kullanılan paketlerden biri olan dplyr paketi, içerdiği fonksiyonlar sayesinde veri düzenleme, sıralama, filtreleme, değişkenleri gruplama, yeni değişkenler oluşturma, seçme ve dönüştürme gibi işlemleri kolaylıkla yapabilmektedir. tidyverse evreninin bir parçası ve sözlük anlamı veri işleme grameri olan bu paket, veri düzenleme ve analizi sürecini oldukça kolaylaştırmaktadır. Veri manipülasyonunun grameri olarak bilinen paketin içerdiği fonksiyonlar, düzgün veri manipülasyonuna olanak tanımaktadır.

Paketin en sık kullanılan fonksiyonları; işlemleri bağlayan bağlama (pipe) %>%, istenilen değişkenlere göre yeni bir veri seti oluşturan select(), gözlemlerin seçilen degişkenlere göre yeniden sıralamaya yarayan arrange(), istenilen gözlemlerle yeni bir veri oluşturan filter(), yeni değişkenleri veri setine ekleyen mutate(), veride grup bazında işlem yapabilen group_by(), veriden özet istatistikleri elde eden summarise() ve verileri birleştirme işlemi yapabilen join() fonsiyonlarıdır. Veri düzenleme I bölümünde select(), filter(), arrange() ve mutate(), veri düzenleme II bölümünde veriyi üst düzeyde toplama işlemlerini gerçekleştiren count(), group_by() ve summarise() fonksiyonları, veri düzenleme III bölümünde birleştirme işlemleri yapan join() fonksiyonları açıklanmıştır. Veri düzenleme IV bölümünde ise veriyi R ortamına aktardıktan sonra verinin analize uygun formata getirilmesine ve tidyr paketi ile birlikte en sık kullanılan fonksiyonlara değinilmiştir.

Veri düzenlemede genellikle birden fazla değişkene ilişkin gözlemlerin yer aldığı veri setleri kullanılır. Pratiklik ve kolay anlaşılabilirlik adına PISA veri setinden değişken sayısının azaltılmasıyla oluşturulan midiPISA veri seti okutulmuş ve gerekli düzenlemeler yapılmıştır.

In [1]:
#Çalışacağımız Veri Setini İnternetten Çekiyoruz.
devtools::install_github("tuevpaket/tuev")
library("tuev")
data(PISA_COG_2018)
data(PISA_SCH_2018)
data(PISA_OGR_2018)

── R CMD build ─────────────────────────────────────────────────────────────────
* checking for file ‘/tmp/RtmpQtRqLt/remotes1d716a677c3/tuevpaket-tuev-92c1d67/DESCRIPTION’ ... OK
* preparing ‘tuev’:
* checking DESCRIPTION meta-information ... OK
Warning in person1(given = given[[i]], family = family[[i]], middle = middle[[i]],  :
  Invalid ORCID iD: ‘YOUR-ORCID-ID’.
* checking for LF line-endings in source and make files and shell scripts
* checking for empty or unneeded directories
* building ‘tuev_0.1.0.tar.gz’



Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)



Veri düzenlemede genellikle birden fazla değişkene ilişkin gözlemlerin yer aldığı veri setleri kullanılır. Pratiklik ve kolay anlaşılabilirlik adına PISA veri setinden değişken sayısının azaltılmasıyla oluşturulan midiPISA veri seti okutulmuş ve gerekli düzenlemeler yapılmıştır.

tidyverse paketi aktif hâle getirildikten sonra oluşturulan data dosyasından midiPISA verisi yüklenir. PISA verileri OECD web adresinden SPSS formatında çekildiği için değişken etiketleri(label) ile birlikte gelmektedir. Bu etiket isimleri bazen R paketlerindeki fonksiyonlar ile birlikte çalışmamaktadır. Bu nedenle expss::drop_var_labs bu etiketlerin kaldırılmasını sağlar ve midiPISA veri seti üzerine kaydedilir.

Bir diğer alternatif ise değişken etiketlerini faktör düzeyi olarak kaydetmektir. Bu işlem aşağıdaki kodlarla sağlanabilir.

In [2]:
#midiPISA veri setimizi oluşturuyoruz
library(dplyr)
library(magrittr)
midiPISA <- PISA_OGR_2018 %>%
  select(OGRENCIID,SINIF,CINSIYET,
         Anne_Egitim,Baba_Egitim,OKUMA_ZEVK,
         ST097Q01TA:ST097Q05TA,ODOKUMA1:ODOKUMA5)


Attaching package: ‘dplyr’


The following objects are masked from ‘package:stats’:

    filter, lag


The following objects are masked from ‘package:base’:

    intersect, setdiff, setequal, union




In [3]:
#miniPISA veri setini oluşturuyoruz.
library(dplyr)
library(magrittr)
miniPISA <- PISA_OGR_2018 %>%
  select(SINIF, CINSIYET, KITAPSAYISI, SES,
    Anne_Egitim,Baba_Egitim,OKUMA_ZEVK,OK_YETERLIK,
    Okuloncesi_yil, OKUL_TUR,ODOKUMA1)

In [76]:
head(miniPISA)

SINIF CINSIYET KITAPSAYISI SES     Anne_Egitim Baba_Egitim OKUMA_ZEVK
1 10    2        2           -2.4521 2           2           -0.2891   
2 10    2        3           -2.1042 2           2            0.6040   
3 10    1        1           -2.2700 1           2            0.6380   
4  9    2        1            0.0324 6           6           -1.1538   
5  9    2        2           -0.0674 4           4            0.6669   
6 10    2        2            0.3978 4           6            0.3568   
  OK_YETERLIK Okuloncesi_yil OKUL_TUR ODOKUMA1
1 -0.6712      1             TUR0112  376.022 
2  1.2374      1             TUR0112  512.316 
3 -0.4089      1             TUR0112  396.383 
4 -0.8250      1             TUR0112  393.006 
5  1.8839     NA             TUR0112  552.457 
6  0.1222     NA             TUR0112  441.286

In [77]:
str(midiPISA)

tibble [6,890 × 16] (S3: tbl_df/tbl/data.frame)
 $ OGRENCIID  : num [1:6890] 79200768 79201064 79201118 79201275 79201481 ...
  ..- attr(*, "label")= chr "OGRENCIID"
  ..- attr(*, "format.spss")= chr "F8.0"
  ..- attr(*, "display_width")= int 10
 $ SINIF      : num [1:6890] 10 10 10 9 9 10 10 10 10 10 ...
  ..- attr(*, "label")= chr "Ogrenci Sinif Duzeyi"
 $ CINSIYET   : num [1:6890] 2 2 1 2 2 2 1 2 1 2 ...
  ..- attr(*, "label")= chr "Ogrenci (Standart) Cinsiyet"
 $ Anne_Egitim: num [1:6890] 2 2 1 6 4 4 5 5 1 0 ...
  ..- attr(*, "label")= chr "Anne Egitim_ISCED"
 $ Baba_Egitim: num [1:6890] 2 2 2 6 4 6 5 5 4 2 ...
  ..- attr(*, "label")= chr "Baba Egitim _ISCED"
 $ OKUMA_ZEVK : num [1:6890] -0.289 0.604 0.638 -1.154 0.667 ...
  ..- attr(*, "label")= chr "Okumaya karsi duyulan zevk"
  ..- attr(*, "format.spss")= chr "F7.4"
  ..- attr(*, "display_width")= int 9
 $ ST097Q01TA : num [1:6890] 1 3 2 2 3 3 3 3 3 3 ...
  ..- attr(*, "label")= chr "ogrenciler ogretmenin soylediklerini dinlemez

**SPSS veya SAS'tan gelen verilerde sütunlar genellikle bir zırh ile kaplıdır. unclass() bu zırhı çıkarıp R'ın altındaki ham veriyi görmesini sağlar.

Veri Setinin Tamamını "Normal" R Formatına Çevirmek
Eğer her seferinde unclass() veya zap_labels() yazmak istemiyorsanız, analizinizin en başında veri setindeki tüm özel etiketleri temizleyip standart R veri tiplerine (sayı veya faktör) dönüştürebilirsiniz.**

İlk başta verdi R hata verdi bu hatayı böyle çözdüm.

In [78]:
# Tüm veri setindeki etiketleri kaldırıp sayıya çevirmek için:
library(haven)
midiPISA <- zap_labels(midiPISA)

Bu kod bloğu, midiPISA veri setiniz üzerinde zincirleme işlemler gerçekleştirmek için boru (%>%) operatörünü kullanır:

`midiPISA:` İşlemlerin başlangıç noktası olan veri setiniz.
`%>%: `Bu operatör, solundaki ifadenin sonucunu sağındaki fonksiyonun ilk argümanı olarak geçirir.
`filter(unclass(SINIF) == 9)`: Bu adım kritik öneme sahiptir. SINIF değişkeni haven_labelled (etiketli) bir veri tipidir. `unclass(SINIF)` fonksiyonu, bu etiketli veri tipinin altında yatan sayısal değerleri ortaya çıkarır, yani `SINIF` değişkenini normal bir sayısal vektöre dönüştürür. Ardından, bu sayısal değerler 9 ile karşılaştırılır ve yalnızca `SINIF` değeri 9 olan satırlar filtrelenir (seçilir).
`head(5):` Son olarak, filtrelenmiş veri setinin ilk 5 satırını görüntülenmesini sağlar. Bu, genellikle veri setinin yapısını ve içeriğini hızlıca incelemek için kullanılır.


# **pipe (%>%)**

%>% simgesiyle ifade edilen pipe (bağlantı) operatörü, aynı veri tabanında yer alan bazı fonksiyonların birlikte tek bir kod hâlinde kullanılmasını sağlamaktadır. Bu durum fazla yazma yükünü ortadan kaldırarak kodu okunabilir kılmaktadır. Böylece veri daha pratik ve anlaşılır hâle gelerek kullanışlılığı artmaktadır. Aslında dplyr paketindeki tüm fonksiyonlar, daha az değişken oluşturmak amacıyla pipe operatörü (%>%) ile kullanılabilir.

%>% operatörü veri düzenleme işlemlerinde her zaman kullanışlı olduğundan sık tercih edilen bir operatördür. Kısa yolu windows için Ctrl+Shift+M, mac için Cmd+Shift+M’dir. Bu operatör magrittr paketinde yer aldığı için pipe operatörü kullanılırken paketin aktifliği kontrol edilmelidir.

%>% solundaki nesneye sağındaki fonksiyonu uygular. Yani pipe operatörünün solundaki öğeler, sağındaki fonksiyonun ilk argümanına iletilir. Fonksiyon ilk argümanı olan veriyi, pipe operatörünün solundan alır, kalan argümanlar fonksiyonun sağındadır.

In [7]:
library(dplyr) # paketin aktifleştirilmesi
library(magrittr) # paketin aktifleştirilmesi
#x %>% f(y) = f(x, y)

In [81]:
#9. sınıflardan ilk 5 kişiyi çağırmak için
midiPISA %>% # veriyi çağırdık
  filter(SINIF == 9) %>%  #9. sınıfları filtreledik
  head(5) #ilk 5 tanesini yazdırdık

OGRENCIID SINIF CINSIYET Anne_Egitim Baba_Egitim OKUMA_ZEVK ST097Q01TA
1 79201275  9     2        6           6           -1.1538    2         
2 79201481  9     2        4           4            0.6669    3         
3 79202354  9     2        4           4           -1.1285    1         
4 79202395  9     2        2           4            1.0123    4         
5 79203125  9     1        5           5            1.3752    3         
  ST097Q02TA ST097Q03TA ST097Q04TA ST097Q05TA ODOKUMA1 ODOKUMA2 ODOKUMA3
1 2          3          1          1          393.006  428.757  364.850 
2 3          4          3          1          552.457  570.109  562.955 
3 1          1          1          1          455.955  551.819  492.381 
4 3          4          3          3          461.661  532.696  467.035 
5 4          3          3          3          418.412  444.452  454.775 
  ODOKUMA4 ODOKUMA5
1 382.522  378.563 
2 530.835  532.487 
3 507.685  510.787 
4 523.916  510.146 
5 415.859  412.982

In [80]:
#Farklı bir syntax ile aynı kodun yazılımı
midiPISA |>
  filter(SINIF == 9) |>
  head(5)

OGRENCIID SINIF CINSIYET Anne_Egitim Baba_Egitim OKUMA_ZEVK ST097Q01TA
1 79201275  9     2        6           6           -1.1538    2         
2 79201481  9     2        4           4            0.6669    3         
3 79202354  9     2        4           4           -1.1285    1         
4 79202395  9     2        2           4            1.0123    4         
5 79203125  9     1        5           5            1.3752    3         
  ST097Q02TA ST097Q03TA ST097Q04TA ST097Q05TA ODOKUMA1 ODOKUMA2 ODOKUMA3
1 2          3          1          1          393.006  428.757  364.850 
2 3          4          3          1          552.457  570.109  562.955 
3 1          1          1          1          455.955  551.819  492.381 
4 3          4          3          3          461.661  532.696  467.035 
5 4          3          3          3          418.412  444.452  454.775 
  ODOKUMA4 ODOKUMA5
1 382.522  378.563 
2 530.835  532.487 
3 507.685  510.787 
4 523.916  510.146 
5 415.859  412.982

# **filter()**

Satır bazında veri seçim işlemi yapmak amacıyla filter() fonksiyonu kullanılır. Böylece veri setindeki istenilen değişkenler filtrelenerek sadece bu gözlemlerden oluşan yeni bir veri seti elde edilir.

filter() fonksiyonu kullanımı;

filter(veri_seti, kosul ve/veya kosullar)

Buradaki filtreleme işlemi tek bir fonsiyon içinde ifade edildiği gibi pipe operatörü ile de kullanılabilir.

filter() fonksiyonunun pipe ile kullanımı;

veri_seti %>% filter(kosul ve/veya kosullar)

In [82]:
kiz <- filter(midiPISA, CINSIYET ==1) %>% # CINSIYETe göre filtreleme
  head(5) #ilk beş satırın görüntülenmesi
kiz

OGRENCIID SINIF CINSIYET Anne_Egitim Baba_Egitim OKUMA_ZEVK ST097Q01TA
1 79201118  10    1        1           2            0.6380    2         
2 79201652  10    1        5           5           -0.0886    3         
3 79202179  10    1        1           4            1.2152    3         
4 79202313  10    1        1           5            1.5806    4         
5 79202360  10    1        4           1            1.5998    2         
  ST097Q02TA ST097Q03TA ST097Q04TA ST097Q05TA ODOKUMA1 ODOKUMA2 ODOKUMA3
1  3         3          3          3          396.383  413.859  423.121 
2 NA         3          3          4          410.648  421.972  425.566 
3  4         3          4          3          541.674  533.529  500.624 
4  4         3          4          4          321.951  312.370  329.122 
5  2         3          2          3          454.225  441.445  441.710 
  ODOKUMA4 ODOKUMA5
1 452.124  392.434 
2 384.666  461.496 
3 522.902  496.677 
4 331.976  340.660 
5 474.351  412.756

In [11]:
midiPISA %>%
  filter(Anne_Egitim==6) %>%
  head(5)

OGRENCIID SINIF CINSIYET Anne_Egitim Baba_Egitim OKUMA_ZEVK ST097Q01TA
1 79201275   9    2        6           6           -1.1538    2         
2 79202343  11    2        6           6           -0.1119    1         
3 79203553  10    1        6           5            1.1914    4         
4 79204714  10    2        6           4            0.3379    3         
5 79200971  10    2        6           5           -0.1672    3         
  ST097Q02TA ST097Q03TA ST097Q04TA ST097Q05TA ODOKUMA1 ODOKUMA2 ODOKUMA3
1 2          3          1          1          393.006  428.757  364.850 
2 2          1          2          2          319.453  333.203  345.495 
3 4          3          2          2          518.787  519.112  502.570 
4 4          4          3          4          335.574  416.818  344.729 
5 2          2          2          2          443.971  427.072  412.744 
  ODOKUMA4 ODOKUMA5
1 382.522  378.563 
2 339.966  346.631 
3 460.873  493.765 
4 350.298  354.592 
5 420.562  439.519

Çıktıda anne eğitim düzeyi sadece lisansüstü(6) olan öğrencilere ait verilerin ilk beş satırı görüntülenmektedir.

In [12]:
midiPISA %>%
  filter(Anne_Egitim==6 & Baba_Egitim==6) %>%
  head(5)

OGRENCIID SINIF CINSIYET Anne_Egitim Baba_Egitim OKUMA_ZEVK ST097Q01TA
1 79201275   9    2        6           6           -1.1538    2         
2 79202343  11    2        6           6           -0.1119    1         
3 79201796  10    2        6           6            0.8418    4         
4 79202928  10    2        6           6           -0.1119    4         
5 79200826  10    1        6           6            2.6131    3         
  ST097Q02TA ST097Q03TA ST097Q04TA ST097Q05TA ODOKUMA1 ODOKUMA2 ODOKUMA3
1 2          3          1          1          393.006  428.757  364.850 
2 2          1          2          2          319.453  333.203  345.495 
3 3          3          4          3          426.008  435.133  453.348 
4 4          4          4          4          337.750  374.678  343.616 
5 3          2          3          3          748.371  660.873  649.216 
  ODOKUMA4 ODOKUMA5
1 382.522  378.563 
2 339.966  346.631 
3 422.474  455.716 
4 319.301  318.048 
5 703.899  683.961

In [13]:
midiPISA %>%
  filter(Anne_Egitim==6 | Baba_Egitim==6) %>%
  head(5)

OGRENCIID SINIF CINSIYET Anne_Egitim Baba_Egitim OKUMA_ZEVK ST097Q01TA
1 79201275   9    2        6           6           -1.1538    2         
2 79201556  10    2        4           6            0.3568    3         
3 79202343  11    2        6           6           -0.1119    1         
4 79203553  10    1        6           5            1.1914    4         
5 79203843  10    2        5           6            0.7796    4         
  ST097Q02TA ST097Q03TA ST097Q04TA ST097Q05TA ODOKUMA1 ODOKUMA2 ODOKUMA3
1 2          3          1          1          393.006  428.757  364.850 
2 3          2          2          3          441.286  415.682  406.586 
3 2          1          2          2          319.453  333.203  345.495 
4 4          3          2          2          518.787  519.112  502.570 
5 4          4          4          4          534.059  524.908  472.845 
  ODOKUMA4 ODOKUMA5
1 382.522  378.563 
2 437.001  473.036 
3 339.966  346.631 
4 460.873  493.765 
5 489.941  447.587

Anne eğitim düzeyi veya baba eğitim düzeyi lisansüstü olan öğrencilerin seçilmesi;

# **select()**

Veri setinden sütun bazında seçim yapmak için select() fonksiyonu kullanılabilir. Bu fonksiyonun kullanımı;

select(veri_seti, degisken_adi, degisken_adi,..)
select() fonksiyonu dplyr içindeki diğer fonksiyonlarla da sıklıkla kullanılır. Örneğin pipe operatörü ile kullanımı;

veri_seti %>% select(degisken_adi, degisken_adi,..)
midiPISA veri seti ile select() fonksiyonunu kullanarak yeni bir veri seti oluşturalım.

In [14]:
library(tidyverse)
midiPISA %>%
  select(OGRENCIID,ST097Q01TA,ST097Q04TA,OKUMA_ZEVK)%>%
head(5)

── Attaching core tidyverse packages ──────────────────────── tidyverse 2.0.0 ──
✔ forcats   1.0.1     ✔ readr     2.1.6
✔ ggplot2   4.0.1     ✔ stringr   1.6.0
✔ lubridate 1.9.4     ✔ tibble    3.3.0
✔ purrr     1.2.0     ✔ tidyr     1.3.1
── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ tidyr::extract()   masks magrittr::extract()
✖ dplyr::filter()    masks stats::filter()
✖ dplyr::lag()       masks stats::lag()
✖ purrr::set_names() masks magrittr::set_names()
ℹ Use the conflicted package (<http://conflicted.r-lib.org/>) to force all conflicts to become errors


OGRENCIID ST097Q01TA ST097Q04TA OKUMA_ZEVK
1 79200768  1          1          -0.2891   
2 79201064  3          3           0.6040   
3 79201118  2          3           0.6380   
4 79201275  2          1          -1.1538   
5 79201481  3          3           0.6669

Görüldüğü üzere dört değişkenden oluşan yeni bir veri seti oluşturulmuştur.select() fonksiyonu, orijinal veri setinde herhangi bir değişikliğe yol açmaz. Bu nedenle seçim yapılan değişkenlerden oluşan yeni veri setini ayrı bir R nesnesine atayarak kaydetmek gereklidir.

Bazen özellikle büyük veri setleri ile çalışıldığında, değişkenlerin tek tek isimlerini yazarak seçmek oldukça vakit alabilir. O nedenle select() fonksiyonu ile belirli bir aralıkta yer alan değişkenler iki nokta : operatörü ile seçilebilirken, bu aralıkta dâhil edilmek istenmeyen değişkenler kısa çizgi - operatörü ile ifade edilebilir.

In [15]:
midiPISA %>%
  select(OGRENCIID:ST097Q04TA,-CINSIYET) %>% # sütun bazında değişken ekleme ve çıkarma
head(5) #ilk beş satırın görüntülenmesi

OGRENCIID SINIF Anne_Egitim Baba_Egitim OKUMA_ZEVK ST097Q01TA ST097Q02TA
1 79200768  10    2           2           -0.2891    1          2         
2 79201064  10    2           2            0.6040    3          2         
3 79201118  10    1           2            0.6380    2          3         
4 79201275   9    6           6           -1.1538    2          2         
5 79201481   9    4           4            0.6669    3          3         
  ST097Q03TA ST097Q04TA
1 1          1         
2 3          3         
3 3          3         
4 3          1         
5 4          3

Elde edilen çıktı incelendiğinde, OGRENCIID, SINIF, Anne_Egitim, Baba_Egitim,OKUMA_ZEVK,ST097Q01TA, ST097Q02TA, ST097Q03TA ve ST097Q04TA değişkenleri seçilmiştir. Bu aralıkta yer alan CINSIYET değişkeni ise seçilmemiştir.

Sadece dplyr paketi içindeki select() fonksiyonu içinde çalışan ve daha detaylı seçimler yapma imkânı sunan bir grup yardımcı fonksiyon içermektedir. Bunlardan en sık kullanılanlar; starts_with(), contains(), matches(), num_range() fonksiyonlarıdır. Burada dikkat edilmesi gereken nokta bu fonksiyonların içindeki ifadenin çift tırnak (“ “) ile beraber kullanılması gerekliliğidir.

starts_with() sadece select() fonksiyonu içinde çalışan ve değişken seçme işlemini kolaylaştıran yardımcı bir fonksiyondur. Örneğin ST097 ile başlayan degişkenler starts_with() fonksiyonu ile seçilebilir.

In [16]:
select(midiPISA, starts_with("ST097")) %>%
  head(5)

ST097Q01TA ST097Q02TA ST097Q03TA ST097Q04TA ST097Q05TA
1 1          2          1          1          1         
2 3          2          3          3          3         
3 2          3          3          3          3         
4 2          2          3          1          1         
5 3          3          4          3          1

In [17]:
# "TA" ile biten degişkenlerin seçimi
select(midiPISA, ends_with("TA")) %>%
  head(5)

ST097Q01TA ST097Q02TA ST097Q03TA ST097Q04TA ST097Q05TA
1 1          2          1          1          1         
2 3          2          3          3          3         
3 2          3          3          3          3         
4 2          2          3          1          1         
5 3          3          4          3          1

Elde edilen çıktı incelendiğinde “TA” ile biten beş değişkenin seçildiği görülmektedir.

contains() yine benzer şekilde içinde belli harfler geçen değişkenler seçilmek istendiğinde kullanılır. Örneğin içinde OD geçen değişkenlerin seçilmesi;

In [18]:
select(midiPISA,contains("OD")) %>%
  head(5)

ODOKUMA1 ODOKUMA2 ODOKUMA3 ODOKUMA4 ODOKUMA5
1 376.022  417.746  420.630  413.837  434.434 
2 512.316  473.229  563.902  485.396  500.394 
3 396.383  413.859  423.121  452.124  392.434 
4 393.006  428.757  364.850  382.522  378.563 
5 552.457  570.109  562.955  530.835  532.487

Elde edilen çıktıya göre içinde “OD” geçen beş okuma olası değerin yer aldığı değişkenler görülmektedir.

matches() içinde 02 geçen değişkenlerin seçilmesinde kullanılır.

In [19]:
select(midiPISA, matches("02")) %>%
  head(5)

ST097Q02TA
1 2         
2 2         
3 3         
4 2         
5 3

Elde edilen çıktıya göre içinde “02” geçen bir değişken görülmektedir.

num_range() fonksiyonu ise ardışık ilerleyen değişkenlerin seçilmesinde kullanılır. midiPISA veri setinde okuma ile ilgili beş olası değer bulunduğundan sadece 1 ve 3 arasındakiler seçilmiştir ve ilk beş gözleme ait değerler istenmiştir. Tabi ki bu örneğimizde değişken sayısı az olduğundan bu seçim çok anlamlı görünmeyebilir fakat özellikle büyük verilerle çalışılırken bu seçim işlemi oldukça pratik olabilmektedir.

In [20]:
select(midiPISA,num_range("ODOKUMA",1:3)) %>%
  head(5)

ODOKUMA1 ODOKUMA2 ODOKUMA3
1 376.022  417.746  420.630 
2 512.316  473.229  563.902 
3 396.383  413.859  423.121 
4 393.006  428.757  364.850 
5 552.457  570.109  562.955

Okuma puanı olası değerleri veri setinde sıralı bir şekilde yer aldığından, bu değerlerden ilk üçü num_range() fonksiyonu kullanılarak seçilmiştir.

Farklı fonksiyonlar ve aynı anda mantıksal operatörler kullanılarak da seçim işlemleri gerçekleştirilebilir. Örneğin veya | operatörü kullanılarak seçim yapılabilir.

midiPISA veri setinde “OD” içeren veya içinde “1” rakamı geçen değişken seçimi;

In [21]:
select(midiPISA,contains("OD") | matches("1")) %>% head(5)

ODOKUMA1 ODOKUMA2 ODOKUMA3 ODOKUMA4 ODOKUMA5 ST097Q01TA
1 376.022  417.746  420.630  413.837  434.434  1         
2 512.316  473.229  563.902  485.396  500.394  3         
3 396.383  413.859  423.121  452.124  392.434  2         
4 393.006  428.757  364.850  382.522  378.563  2         
5 552.457  570.109  562.955  530.835  532.487  3

Elde edilen çıktı incelendiğinde, içinde “OD” veya “1” geçen toplam altı değişken olduğu görülmektedir. & operatörünün kullanımı da oldukça fazladır.

midiPISA veri setinde A harfini içeren ve içinde “2” rakamı olan değişkenlerin seçimi için;

In [22]:
select(midiPISA,contains("A") & matches("2")) %>% head(5)

ST097Q02TA ODOKUMA2
1 2          417.746 
2 2          473.229 
3 3          413.859 
4 2          428.757 
5 3          570.109

Elde edilen çıktı incelendiğinde, midiPISA veri setinden “O” içermeyen ve içinde “0” rakamı olmayan SINIF, CINSIYET, Anne_Egitim ve Baba_Egitim değişkenleri seçilmiştir.

# **arrange()**

arrange() fonksiyonu satırları/gözlemleri sıralamak amacıyla kullanılır. Sıralama işlemini alfabetik sıralamaya göre yapar. Fakat ek bir fonksiyon yazıldığında büyüklük sırasına göre de sıralama yapar. Böylece veri seti istenilen kritere göre sıralanarak değişkenler arasındaki ilişki daha pratik bir şekilde görülebilir. arrange() fonksiyonunun kullanımı oldukça kolaydır; fonksiyonun içine veri setinin adını ve sıralama için kullanılacak değişkeni yazmak yeterli olur.

arrange() fonksiyonunun kullanımı;

arrange() fonksiyonunun kullanımı;
arrange(veri_seti, degisken_adi)
arrange() fonksiyonunun pipe operatörü %>% ile kullanımı;

veri_seti %>% arrange(degisken_adi)

Örnek veri seti üzerinden dört değişken seçip yeni bir veri setine (df) atayarak kaydedelim.

In [23]:
df <- midiPISA %>%
  select(OGRENCIID,ST097Q01TA,ST097Q04TA,OKUMA_ZEVK)
  df %>% head(6)

OGRENCIID ST097Q01TA ST097Q04TA OKUMA_ZEVK
1 79200768  1          1          -0.2891   
2 79201064  3          3           0.6040   
3 79201118  2          3           0.6380   
4 79201275  2          1          -1.1538   
5 79201481  3          3           0.6669   
6 79201556  3          2           0.3568

In [24]:
#Yeni oluşan df veri setini OKUMA_ZEVK(okumaktan zevk alma) puanlarına göre sıralayalım.
arrange(df,OKUMA_ZEVK) %>%
  head(5)

OGRENCIID ST097Q01TA ST097Q04TA OKUMA_ZEVK
1 79204460  1          1          -2.7316   
2 79201124  1          1          -2.7114   
3 79204401  3          2          -2.7114   
4 79206724  1          3          -2.7114   
5 79204126  4          4          -2.7114

Elde edilen çıktı incelendiğinde, arrange() fonksiyonu ile öğrencilerden elde edilen gözlemler okumaktan zevk alma puanlarına göre küçükten büyüğe sıralanmıştır. Çünkü paket içerisinde yer alan fonksiyon sıralamayı önceden tanımlı/default olarak küçükten büyüğe yapar. Fakat bu gözlemler, büyükten küçüğe sıralanmak istendiğinde ek olarak arrenge() fonksiyonu içine desc() fonksiyonu yazılır.

In [25]:
arrange(df,desc(OKUMA_ZEVK))%>%  #büyükten küçüğe sıralama
  head(5)

OGRENCIID ST097Q01TA ST097Q04TA OKUMA_ZEVK
1 79206987  3          2          2.6574    
2 79204432  3          3          2.6574    
3 79206337  4          4          2.6574    
4 79203300  3          2          2.6131    
5 79205358  4          2          2.6131

In [26]:
midiPISA %>%
 select(OGRENCIID,ST097Q01TA,ST097Q04TA,OKUMA_ZEVK) %>% #değişkenlerin seçimi
  arrange(OKUMA_ZEVK)%>% # değişkendeki gözlemleri sıralama
  head(6) #ilk 6 satırın görüntülenmesi

OGRENCIID ST097Q01TA ST097Q04TA OKUMA_ZEVK
1 79204460  1          1          -2.7316   
2 79201124  1          1          -2.7114   
3 79204401  3          2          -2.7114   
4 79206724  1          3          -2.7114   
5 79204126  4          4          -2.7114   
6 79205685  3          3          -2.7114

# **rename()**

Veri setinden istenilen değişkenler seçilerek yeni bir veri seti oluşturulmak istendiğinde seçilen değişkenlerin de ismini değiştirmek isteyebiliriz. Bu durumda rename() fonksiyonu (yeni ad=eski ad) şeklinde kullanılabilir.

In [27]:
midiPISA %>%
select(ODOKUMA1,ODOKUMA2)%>%
  rename(okumapuan1=ODOKUMA1,okumapuan2=ODOKUMA2) %>%
head(3)

okumapuan1 okumapuan2
1 376.022    417.746   
2 512.316    473.229   
3 396.383    413.859

Görüldüğü üzere veri setinden seçilen ODOKUMA1 ve ODOKUMA2 sütunlarının isimleri okumapuan1 ve okumapuan2 olarak değiştirilmiştir.

select() fonksiyonu içinde değişken seçimi sırasında da değişken adı değişimi yapılabilir. Örnekte midiPISA veri setinden dört değişken seçilerek oluşturulan yeni veri setinde okuma becerileri olası değerleri 1 in yer aldığı ODOKUMA1 değişkeninin adı okumapuan1 olarak değiştirilip seçilebilir.

In [28]:
midiPISA %>%
select(OGRENCIID,ST097Q02TA,ST097Q03TA, okumapuan1=ODOKUMA1) %>%
head(3)

OGRENCIID ST097Q02TA ST097Q03TA okumapuan1
1 79200768  2          1          376.022   
2 79201064  2          3          512.316   
3 79201118  3          3          396.383

Yeni seçilen veri setinde sadece ODOKUMA1 değişkeninin ismi okumapuan1 olarak değiştirilmiştir. Diğer üç değişkenin isminin aynı kaldığı görülmektedir.

# **mutate()**

mutate() fonksiyonu ile veri setine yeni değişkenler eklenirken mevcut değişkenler de korunur ancak transmute() fonksiyonu ile eski değişkenler veri setiden çıkarılarak yeni değişken eklenir.

mutate() fonksiyonunun kullanımı;

mutate(veri_seti, yeni_değişken = değişken1 + değişken2)
Örneğin ST097 ile başlayan maddelerden yeni bir veri seti oluşturulup, bu veri setine toplam puan eklenebilir. Aşağıdaki örnekte ST097 ile başlayan maddeler seçilmiş ve seçilen beş maddenin toplamı ayrı bir değişken olarak eklenmiştir. Oluşturulan bu toplam puan değişkeni ve beş madde zevk adlı yeni bir veri seti olarak eklenmiştr.

In [29]:
zevk<- select(midiPISA, starts_with("ST097"))
zevk%>%
mutate(toplam =ST097Q01TA+ST097Q02TA+ST097Q03TA+ST097Q04TA+ST097Q05TA) %>%
 head(3)

ST097Q01TA ST097Q02TA ST097Q03TA ST097Q04TA ST097Q05TA toplam
1 1          2          1          1          1           6    
2 3          2          3          3          3          14    
3 2          3          3          3          3          14

Bu örnekte değişken değerleri + ile toplanacağı gibi rowSums() fonksiyonu ile de aşağıdaki gibi toplanabilir. Bu bize özellikle çok fazla maddenin yer aldığı ölçeklerde toplam puan almada pratiklik sağlar. Örnekte kullanılan across(), aynı dönüşümü birden fazla sütuna uygulamayı kolaylaştırarak summarise() ve mutate() gibi fonkisyonların içinde select() ile aynı işlevi kullanmanıza olanak tanır.

In [30]:
zevk %>%
mutate(toplam=rowSums(across(ST097Q01TA:ST097Q05TA))) %>%
head(3)

ST097Q01TA ST097Q02TA ST097Q03TA ST097Q04TA ST097Q05TA toplam
1 1          2          1          1          1           6    
2 3          2          3          3          3          14    
3 2          3          3          3          3          14

Ayrıca yeni değişken üretiminde sadece toplama değil istenen her türlü işlem yapılabilir. Yeni eklenecek yerler .before ya da .after argümanları ile düzenlenebilir.

midiPISA veri setinde yer alan ST097Q01TA, ST097Q02TA, ST097Q03TA, ST097Q04TA ve ST097Q05TA maddelerinden elde edilen toplam puan değişkeni “ST097Q01TA” değişkeninin hemen önüne eklenir.

In [31]:
zevk %>%
mutate(toplam=rowSums(across(ST097Q01TA:ST097Q05TA)),
       .before= ST097Q01TA)%>%
  head(3)

toplam ST097Q01TA ST097Q02TA ST097Q03TA ST097Q04TA ST097Q05TA
1  6     1          2          1          1          1         
2 14     3          2          3          3          3         
3 14     2          3          3          3          3

Ek olarak mutate() fonksiyonu ile birden fazla değişken de eklenebilir. Yani tek bir kod ile birden fazla yeni değişken üretilebilir.

mutate(veri_seti, yeni_değişken1 = değişken1 + değişken2, yeni_değişken2 = değişken3 * değişken4)

# **transmute()**

transmute() fonksiyonu kullanılarak yeni değişken oluşturulduğunda, işlemde kullanılan değişkenler veri setinden çıkarılır, sadece yeni oluşturulan değişken veri setine eklenir.

In [32]:
zevk %>%
transmute(toplam=rowSums(across(ST097Q01TA:ST097Q05TA))) %>%
head(2)

toplam
1  6    
2 14

# **ifelse()**

ifelse() programlama dillerinde sıklıkla kullanılan koşullu önermelerden biridir. ifelse() fonksiyonunun kullanımı;

ifelse(test = x<0, evet = ilkdeger , hayır = ikincideger)
Bir örnek üzerinde ifelse() fonksiyonunu açıklayalım. Öncelikle x nesnesini aşağıdaki gibi tamsayılardan oluşturalım ve ifelse() fonksiyonu sayesinde sıfırdan küçük olanlar negatif, büyük olanlar pozitif olacak şekilde tanımlansın.

In [33]:
x <- c(-2,1,-1,-3,3)
ifelse(x<0,"Negatif", "Pozitif")

[1] "Negatif" "Pozitif" "Negatif" "Negatif" "Pozitif"

In [34]:
Okultur <- midiPISA %>%
  select(1:5) %>%  #ilk beş değişkenin seçimi
  mutate(okul = ifelse(SINIF == 7 | SINIF == 8,
                       "Ortaokul", "Lise")) %>%  # okul değişkeninin veri setine eklenmesi
 arrange(SINIF) # veri setinin SINIF değişkenine göre sıralanması

tail(Okultur)

OGRENCIID SINIF CINSIYET Anne_Egitim Baba_Egitim okul
1 79203092  12    1        0           1           Lise
2 79204140  12    2        0           1           Lise
3 79200174  12    1        5           5           Lise
4 79206519  12    1        0           0           Lise
5 79205173  12    1        5           5           Lise
6 79201478  12    1        0           1           Lise

In [35]:
head(Okultur,3)

OGRENCIID SINIF CINSIYET Anne_Egitim Baba_Egitim okul    
1 79204097  7     1         2           2          Ortaokul
2 79207110  7     2         6           1          Ortaokul
3 79202600  7     2        NA          NA          Ortaokul

Elde edilen çıktıda ilk üç satırdaki sınıf düzeyleri 7 olduğundan okul değişkeninin ilk üç satırı da ortaokul olarak kategorilendirilmiştir.

# **case_when()**

case_when() fonksiyonu çoklu ifelse() kullanımı ile benzer işlevi sağlar. case_when() birden fazla koşula dayalı karşılaştırmalarda yeni bir değişken oluşturmak amacıyla kullanılır.

ODOKUMA1 düzeyine ilişikin betimsel istatistikler;

In [36]:
summary(midiPISA$ODOKUMA1)

   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
  175.6   402.6   463.4   464.2   525.7   771.5 

ODOKUMA1 değişkeninin kategorik hale getirilmesinde çeyreklik puanları kullanılmıştır.

In [37]:
v1 <- midiPISA %>%
  mutate(ODOKUMA1_kategorik =
    case_when(
      ODOKUMA1 <=  402.6  ~ "dusuk",
      ODOKUMA1 > 402.6  & ODOKUMA1 <  525.7 ~ "orta",
      ODOKUMA1 >=525.7 ~ "yuksek" )) %>%
      select(ODOKUMA1,ODOKUMA1_kategorik)
  head(v1)

ODOKUMA1 ODOKUMA1_kategorik
1 376.022  dusuk             
2 512.316  orta              
3 396.383  dusuk             
4 393.006  dusuk             
5 552.457  yuksek            
6 441.286  orta

Okuma puanı 402.6’dan küçük olanlar düşük, 402.6 ile 525.7 arasında olanlar orta, 525.7 den büyük olanlar ise yüksek olarak kategorilendirilmiştir. Oluşturulan ODOKUMA1_kategorik adlı yeni değişken ile okuma puanı olası değerlerinin yer aldığı ODOKUMA1 değişkeninden oluşan yeni veri seti v1 olarak çalışma alanına kaydedilmiştir. head() fonksiyonu ile v1 veri setinin ilk altı satırı görüntülenmektedir.


Veri düzenlemenin diğer bir aşaması ise veriyi üst düzeyde toplama işlemleridir. Bu işlemlerden en sık kullanılanlar; count(), grup_by(), summarise(), summarise_all(), summarise_if(), summarise_at(),across(), top_n() fonksiyonlarıdır. Yine bu fonksiyonlar dplyr paketi içerisinde yer almakatadır. Sırasıyla örneklerle bu fonksiyonları inceleyelim.

# **count() fonksiyonu**

count() fonksiyonu frekans tablosu oluşturmak için kullanılmaktadır. Yine pipe operatörü ile birlikte sıklıkla kullanılmaktadır.

count() fonksiyonunun pipe ile kullanımı:

veri_seti %>% count(degisken_adi)
Görüldüğü üzere fonksiyonun içine değişken ismini yazmak yeterlidir. Örnek veri seti üzerinden fonksiyonun kullanımını inceleyelim.

In [83]:
midiPISA %>% count() # frekans hesaplanması

count(midiPISA) #aynı iş farklı syntax

n   
1 6890

n   
1 6890

Elde edilen çıktıya göre midiPSA verisinde 6890(n) gözlem olduğu görülmektedir. Fakat sadece tek bir değişkene göre dağılım incelenmek istenirse fonksiyonun içine değişken ismi yazılmalıdır.

In [92]:
# CINSIYET değişkenine frekans hesaplama
midiPISA |> count(CINSIYET)



CINSIYET n   
1 1        3396
2 2        3494

Çıktıda cinsiyet değişkenine göre frekans tablosu yer almaktadır. Kız(1) öğrencilerin sayısı 3396, erkek öğrencilerin sayısı 3494 olduğu görülmektedir. Ayrıca cinsiyete göre dağılımlar sıralanmak istenebilir. Bu durumda sort argümanı ile kullanılabilir.

In [40]:
# CINSIYET değişkenine göre frekansların sıralanması
midiPISA %>% count(CINSIYET,sort=TRUE)

CINSIYET n   
1 2        3494
2 1        3396

Elde edilen çıktıda sort argümanı yardımıyla frekans tablosundaki değerlerin sıralandığı görülmektedir. Şimdi bu veri setini kullanarak öğrencilerin SINIF ve CINSIYET değişkenlerine göre dağılımını sıralama argümanını kullanarak inceleyelim.

In [41]:
# CINSIYET ve SINIF değişkenlerine göre dağılımın sıralanması
midiPISA %>% count(CINSIYET,SINIF, sort=TRUE)

CINSIYET SINIF n   
1  1        10    2707
2  2        10    2653
3  2         9     747
4  1         9     548
5  1        11     124
6  2        11      83
7  1         8      11
8  2         8       8
9  1        12       5
10 2         7       2
11 1         7       1
12 2        12       1

Elde edilen çıktıda sınıf düzeyi (SINIF), cinsiyet (CINSIYET) ve frekansların (n) yer aldığı üç sütunun olduğu görülmektedir. “sort=TRUE” yardımıyla n sütunundaki değerler büyükten küçüğe sıralanarak verilmiştir. CINSIYET değişkeni iki kategori, SINIF değişkeni altı kategoriden oluştuğundan toplam 12 satır yer almaktadır. Frekansı en yüksek olan grup (n=2707) 10. sınıfa ait kız öğrencilerdir(1).

# **summarise()/summarize()**

summarise() fonksiyonu tek satırda veri setini özetleyerek yeni bir veri seti oluşturan fonksiyondur. Yani seçilen sütunlar için her satırı kullanarak özet istatistikleri hesaplar. Örneğin; min() minumum değer, max() maksimum değer, mean() ortalama değer, median() ortanca, quantile() q. yuzdelik, sd() standart sapma, var() varyans, diff(range())değiskenlik, first() ilk eleman, last() son eleman, nth() n. eleman n() toplam eleman sayısı, n_distinct() farklı değerlerin sayısı hesaplanabilir.

midiPISA veri setinde yer alan okuma başarısı için hesaplanan olası puan değerlerinden (plausible value) ilki kullanılmıştır. Burada okuma puanlarının ortalaması hesaplanabilir.

In [42]:
midiPISA %>%
summarise(mean(ODOKUMA1)) #ortalamanın hesaplanması

mean(ODOKUMA1)
1 464.2299

Genellikle bir veri setinde özet bilgiler elde etmek için birden fazla özetleyici fonksiyon kullanılmak istenebilir. Örneğin bir veri setindeki değişkenlerin hem ortalaması, hem standart sapması hem de minimum, maksimum değerleri hesaplanmak istenebilir. Bu durumda aşağıdaki örnekte olduğu gibi bu fonksiyonlar “,” kullanılarak art arda yazılabilir.

In [43]:
midiPISA %>%
  summarise(n = n(),
            ortalama=mean(ODOKUMA1),
            sd=sd(ODOKUMA1),
            min=min(ODOKUMA1),
            max=max(ODOKUMA1),
            ortanca=quantile(ODOKUMA1),
            first(ODOKUMA1),
            last(ODOKUMA1),
            n_distinct(ODOKUMA1)
            )

Warning message:
“Returning more (or less) than 1 row per `summarise()` group was deprecated in
dplyr 1.1.0.
ℹ Please use `reframe()` instead.
ℹ When switching from `summarise()` to `reframe()`, remember that `reframe()`
  always returns an ungrouped data frame and adjust accordingly.”


n    ortalama sd       min     max     ortanca  first(ODOKUMA1)
0%   6890 464.2299 87.78006 175.608 771.508 175.6080 376.022        
25%  6890 464.2299 87.78006 175.608 771.508 402.5635 376.022        
50%  6890 464.2299 87.78006 175.608 771.508 463.4030 376.022        
75%  6890 464.2299 87.78006 175.608 771.508 525.7188 376.022        
100% 6890 464.2299 87.78006 175.608 771.508 771.5080 376.022        
     last(ODOKUMA1) n_distinct(ODOKUMA1)
0%   415.541        6834                
25%  415.541        6834                
50%  415.541        6834                
75%  415.541        6834                
100% 415.541        6834

# **summarise() ve group_by()**

summarise() fonksiyonu gruplandırılmamış bir veri setinde, tüm satırlardan özet istatistik bilgileri hesaplamıştır. Bu bilgiler, veri setinde yer alan alt gruplar için ise ayrı ayrı group_by() fonksiyonu ile hesaplanabilir. group_by() dan sonra kullanılan fonksiyonlar her grup için ayrı ayrı hesaplanama yapar ve bu fonksiyon içinde sürekli değişken kullanılmaz.

CINSIYET değişkenine göre ODOKUMA1 puanlarına ilişkin özetleyici istatistikler:

In [44]:
midiPISA %>%
  group_by(CINSIYET) %>%
  summarise(n = n(),ortalama=mean(ODOKUMA1),sd=sd(ODOKUMA1),min=min(ODOKUMA1),max=max(ODOKUMA1))

CINSIYET n    ortalama sd       min     max    
1 1        3396 478.1378 83.65322 236.389 771.508
2 2        3494 450.7121 89.57864 175.608 747.491

Elde edilen çıktıda, CINSIYET değişkenine göre gruplandırma yapılarak kız ve erkekler için frekans, ortalama, standart sapma, minimum ve maksimum değerler hesaplanmıştır.

Özetleyici istatistiksel bilgiler, veri setinde yer alan birden fazla kategorik değişken için de hesaplanabilir. Öğrencilerin cinsiyet ve sınıf düzeylerine göre elde edilen betimsel istatistikleri ortalamaya göre büyükten küçüğe sıralanmıştır. Yapılan bu işlem “betimsel” isimli nesneye atanmıştır.

In [45]:
betimsel <- midiPISA%>%  # betimsel veri nesnesine atama
# CINSIYET ve SINIF değişkenlerine göre gruplara ayırma
  group_by(CINSIYET,SINIF) %>%
# özetleyici bilgileri hesaplama
  summarise(n = n(),ortalama=mean(ODOKUMA1),sd=sd(ODOKUMA1)) %>%
# gözlemleri ortalama değerleri büyükten küçüğe olacak şekilde sıralama
  arrange(desc(ortalama))
head(betimsel)

`summarise()` has grouped output by 'CINSIYET'. You can override using the
`.groups` argument.


CINSIYET SINIF n    ortalama sd      
1 1        10    2707 482.2966 79.87708
2 1        11     124 472.6618 84.95710
3 1         9     548 462.0539 96.90558
4 2        10    2653 459.2322 85.00155
5 2        11      83 447.9737 87.87190
6 2         9     747 422.1915 98.74913

Çıktıda olduğu gibi veri seti SINIF ve CINSIYET değişkenlerine göre gruplandırılarak ortalama değerler büyükten küçüğe sıralı olacak şekilde elde edilmiştir. Örneğin en yüksek ortalamaya sahip (482.2971) olan erkek ve 10. sınıftan 2707 öğrenci vardır. group_by() fonksiyonu ile elde ettiğiniz çıktılarda aşağıdaki gibi gruplandırılmış veri Groups çıktısı ile alınır.

Gruplandırılmış elde edilen veri setlerinde tekrar işlem yapmak istenirse ungroup() fonksiyonu kullanılabilir.

In [46]:
midiPISA %>%
  #CINSIYET ve SINIF değişkenlerine göre gruplara ayırma
  group_by(CINSIYET,SINIF) %>%
  # özetleyici bilgileri hesaplama
  summarise(n = n(),ortalama=mean(ODOKUMA1),sd=sd(ODOKUMA1)) %>%
  # gözlemleri ortalama değerleri büyükten küçüğe olacak şekilde sıralama
  arrange(desc(ortalama)) %>%
  # grupları birleştirme
  ungroup()

`summarise()` has grouped output by 'CINSIYET'. You can override using the
`.groups` argument.


CINSIYET SINIF n    ortalama sd      
1  1        10    2707 482.2966 79.87708
2  1        11     124 472.6618 84.95710
3  1         9     548 462.0539 96.90558
4  2        10    2653 459.2322 85.00155
5  2        11      83 447.9737 87.87190
6  2         9     747 422.1915 98.74913
7  1        12       5 421.8288 96.60689
8  2         8       8 362.9183 82.82819
9  1         8      11 355.5180 62.47101
10 1         7       1 343.5010       NA
11 2         7       2 330.3725 62.13135
12 2        12       1 322.0710       NA

Gruplandırılmış veri setlerinde bazı fonksiyonlar çalışmayabileceği için, üretilen yeni veri seti başka amaçlarla kullanılacağında ungroup() fonksiyonu ile grupları birleştirmek gereklidir.

# **across()**

Bir veri setinde aynı anda birden fazla sütuna aynı işlem uygulanmak istendiğinde dplyr paketi içindeki across() fonksiyonu sıklıkla kullanılmaktadır. Bu fonksiyon veri düzenleme ile ilgili birçok temel fonksiyon içinde düzgün çalışabilmektedir. Fakat genellikle select(), mutate(), filter() veya summarise() içinde kullanılır.

cols = argümanına sütunlar ve .fns = argümanına uygulanacak fonksiyonlar atanır.

midiPISA verisinde okuma puanı olası değer 1 ve 2 sütunlarına ait ortalama değerleri across() fonksiyonu ile hesaplayalım.

In [47]:
midiPISA %>%
     summarise(across(.cols=c(ODOKUMA1,ODOKUMA2),.fns=mean, .names = "{col}_mean"))

ODOKUMA1_mean ODOKUMA2_mean
1 464.2299      464.4204

ODOKUMA1 ve ODOKUMA2 sütunlarına ait ortalamalar hesaplanmış ve bu ortalama değerlerini veren sütunlar .names = “{col}_mean” ile isimlendirilmiştir.

Birden fazla istatistiksel bilgi hesaplanmak istendiğinde list() argümanı kullanılabilir. “OD” ile başlayan sütunlara ait ortalama ve standart sapma değerlerini hesaplayalım.

In [48]:
midiPISA %>%
       summarise(across(.cols=starts_with("OD"), .fns=list(mean = mean, sd = sd)))

ODOKUMA1_mean ODOKUMA1_sd ODOKUMA2_mean ODOKUMA2_sd ODOKUMA3_mean ODOKUMA3_sd
1 464.2299      87.78006    464.4204      87.696      464.7147      87.07692   
  ODOKUMA4_mean ODOKUMA4_sd ODOKUMA5_mean ODOKUMA5_sd
1 464.6129      87.40305    464.1982      87.21323

# **summarise() ve across()**

dplyr paket fonksiyonlarının **_at,_if,_all** uzantılı varyasyonları bulunmaktadır. Bunlardan biri olan summarise_at() fonksiyonunu bir grup sütunun ortalamasını ve standart sapmasını hesaplamak gerektiğinde kullanabilirsiniz. summarise_at() fonksiyonu ile seçilecek değişkenler vars() fonksiyonu içinde belirtilebilir. Bu işlem select() işlemi yerine geçmektedir. Hesaplama işlemlerini ise list() fonksiyonu içinde tanımlayabilirsiniz.

In [49]:
midiPISA %>%
    summarise_at(vars(ODOKUMA1, ODOKUMA2), list(~mean(.), ~sd(.)))

ODOKUMA1_mean ODOKUMA2_mean ODOKUMA1_sd ODOKUMA2_sd
1 464.2299      464.4204      87.78006    87.696

In [50]:
#Aynı işlemin farklı bir syntax ile yapılması
midiPISA %>%
  summarise(
    across(c(ODOKUMA1, ODOKUMA2), list(mean = ~mean(.), sd = ~sd(.)))
  )

ODOKUMA1_mean ODOKUMA1_sd ODOKUMA2_mean ODOKUMA2_sd
1 464.2299      87.78006    464.4204      87.696

In [51]:
#summarise_all () fonksiyonu ile tüm sütunlara istenilen fonksiyon uygulanabilir.
midiPISA %>%
  summarise(across(everything(), list(mean = mean, sd = sd),na.rm=TRUE))

Warning message:
“There was 1 warning in `summarise()`.
ℹ In argument: `across(everything(), list(mean = mean, sd = sd), na.rm =
  TRUE)`.
Caused by warning:
! The `...` argument of `across()` is deprecated as of dplyr 1.1.0.
Supply arguments directly to `.fns` through an anonymous function instead.

  # Previously
  across(a:b, mean, na.rm = TRUE)

  # Now
  across(a:b, \(x) mean(x, na.rm = TRUE))”


OGRENCIID_mean OGRENCIID_sd SINIF_mean SINIF_sd  CINSIYET_mean CINSIYET_sd
1 79203623       2086.54      9.83701    0.4581423 1.507112      0.4999857  
  Anne_Egitim_mean Anne_Egitim_sd Baba_Egitim_mean Baba_Egitim_sd ⋯
1 2.655157         1.951497       3.124835         1.941041       ⋯
  ODOKUMA1_mean ODOKUMA1_sd ODOKUMA2_mean ODOKUMA2_sd ODOKUMA3_mean ODOKUMA3_sd
1 464.2299      87.78006    464.4204      87.696      464.7147      87.07692   
  ODOKUMA4_mean ODOKUMA4_sd ODOKUMA5_mean ODOKUMA5_sd
1 464.6129      87.40305    464.1982      87.21323

Çıktıda tüm değişkenlerin eksik veriler silinerek ortalamasının alındığı görülmektedir.

Elinizdeki bir veri setinin sayısal (numeric) olan sütunlarının ortalamasını summarise_if() fonksiyonu ile hesaplayabilirsiniz. Bu hesaplamanın summarise() ve across() fonksiyonu ile nasıl yapıldığı gösterilmiştir.

In [52]:
midiPISA %>%
  summarise(across(where(is.numeric), list(mean = mean, sd = sd), na.rm = TRUE))

OGRENCIID_mean OGRENCIID_sd SINIF_mean SINIF_sd  CINSIYET_mean CINSIYET_sd
1 79203623       2086.54      9.83701    0.4581423 1.507112      0.4999857  
  Anne_Egitim_mean Anne_Egitim_sd Baba_Egitim_mean Baba_Egitim_sd ⋯
1 2.655157         1.951497       3.124835         1.941041       ⋯
  ODOKUMA1_mean ODOKUMA1_sd ODOKUMA2_mean ODOKUMA2_sd ODOKUMA3_mean ODOKUMA3_sd
1 464.2299      87.78006    464.4204      87.696      464.7147      87.07692   
  ODOKUMA4_mean ODOKUMA4_sd ODOKUMA5_mean ODOKUMA5_sd
1 464.6129      87.40305    464.1982      87.21323

Mevcut kodunuzu _if, _at veya _all işlevleri yerine across() işlevini kullanacak şekilde güncellemek istiyorsanız summarise_at(), summarise_all() ve summarise_if() fonksiyonlarının yerini summarise() fonksiyonu içinde de kullanılabilen across() yardımcı fonksiyonunu kullanabilirsiniz.

# **top_n()**

top_n() fonksiyonu ile istediğiniz bir değişkenin en yüksek ya da en düşük değerlerine göre veri setinde seçim yapılabilir.

In [53]:
df <- data.frame(x = c(10, 4, 1, 6, 3, 1, 1))
df %>% top_n(2)

Selecting by x


x 
1 10
2  6

Okuma puanı en yüksek olan beş kız ve beş erkek öğrencilerin bilgileri

In [54]:
midiPISA %>%
  # CINSIYET ve okuma olası değer1 değişkenlerinin seçilmesi
  select(CINSIYET,ODOKUMA1)%>%
  # büyükten küçüğe okuma puanlarının sıralanması
  arrange(desc(ODOKUMA1))%>%
  # CINSIYET değişkenine göre verinin gruplandırılması
  group_by(CINSIYET) %>%
  # okuma puanına göre her kategoriye ait en yüksek 5'er öğrencinin görüntülenmesi
  top_n(5,ODOKUMA1)

CINSIYET ODOKUMA1
1  1        771.508 
2  1        748.371 
3  2        747.491 
4  1        742.969 
5  2        737.448 
6  1        718.849 
7  1        714.855 
8  2        713.531 
9  2        713.347 
10 2        706.751

# **top_n() & - operatörü**

top_n() fonksiyonu, “-” ile birlikte kullanıldığında, veri setindeki en düşük ilgili özelliğe sahip öğrencilerin bilgilerini verir. Örneğin, okuma puanı en düşük olan beş kız ve beş erkek öğrencinin bilgilerini elde edelim:

In [93]:
midiPISA %>%
  select(CINSIYET,ODOKUMA1)%>%
  arrange(desc(ODOKUMA1))%>%
  group_by(CINSIYET) %>%
  # okuma puanına göre her kategoriye ait en düşük 5'er öğrencinin görüntülenmesi
  top_n(-5,ODOKUMA1)

CINSIYET ODOKUMA1
1  1        254.300 
2  1        253.454 
3  1        249.789 
4  1        241.820 
5  1        236.389 
6  2        220.286 
7  2        210.915 
8  2        198.944 
9  2        176.962 
10 2        175.608

Elde edilen çıktı incelendiğinde, okuma puanı en düşük 5 öğrenci hem erkek hem de kız öğrenciler olarak ayrı ayrı belirlenmiştir.

Bu bölümde veriyi bir üst düzeye toplama ile ilişki fonksiyonlara yer verilmiştir. Veri Düzenleme III bölümünde farklı veri setlerini birleştirme işlemlerine değinilecektir.

# **join()**

join() fonksiyonları iki veri setini istenilen şekilde birleştirme amacıyla kullanılır. Örneğin elimizde A ve B olmak üzere iki farklı veri seti olsun. Her iki veri setini birleştirmek istediğimizde bu veri setlerinden hangi satır veya sütunları seçeceğimizi, satırların eşleşip eşleşmeyeceğini hangi değişkenlerle belirleneceğinin bilinmesi gerekmektedir. Bu nedenle de her bir amaca yönelik join fonksiyon türleri belirlenmiştir. Bunlar; left_join(), right_join(), full_join(), inner_join(), semi_join(), anti_join() fonksiyonlarıdır.

# **left_join()**

A %>% left_join(B) ile A verisindeki tüm satırlar, mümkün olduğunda B verisi ile eşleştirilerek (olmadığında “NA” verir), hem A hem de B den gelen sütunlar alınır.

left_join() fonksiyonu ile öncelikle X1 değişkenine ait tüm gözlemler alınmıştır. X2 değişkenine ait ID numarası “1” olan gözlem olmadığı için yeni veri setinde bu kısım kayıp veri (NA) olarak girilmiştir. Burada birleştirme yapılacak iki veri setine ait örnekteki gibi ortak bir değişkenin (“ID”) olması önemlidir.

Örnek bir veri seti üzerinden açıklayalım. midiPISA1, midiPISA verisinden OGRENCIID, CINSIYET ve ST097Q01TA, ST097Q02TA değişkenlerinin ve bu veri setinin ilk altı satırının seçilmesi ile oluşturulur.

In [56]:
midiPISA1 <- midiPISA %>% select(OGRENCIID,CINSIYET,ST097Q01TA,ST097Q02TA) #değişkenlerin seçimi
midiPISA1 <-midiPISA1[1:6,] #veri setinin ilk 6 satırının seçilmesi
midiPISA1

OGRENCIID CINSIYET ST097Q01TA ST097Q02TA
1 79200768  2        1          2         
2 79201064  2        3          2         
3 79201118  1        2          3         
4 79201275  2        2          2         
5 79201481  2        3          3         
6 79201556  2        3          3

midiPISA1 verisinin oluşturulmasından sonra midiPISA2 verisi oluşturulur. Öncelikle midiPISA verisinden öğrenci id, okumaktan zevk alma ve okuma olası değer1 puanları değişkenleri seçilir. Ardından bu veri setinin ilk yedi satırı seçilir ve üçüncü satır silinir. Sonuç olarak altı satırlık bir midiPISA2 veri seti elde edilmiş olur.

In [57]:
#değişkenlerin seçimi
midiPISA2<- midiPISA %>% select(OGRENCIID,OKUMA_ZEVK,ODOKUMA1)
midiPISA2<-midiPISA2[1:7,] #veri setinin ilk 7 satırının seçilmesi
midiPISA2<-midiPISA2[-3,] #veri setinin 3. satırının çıkarılması
midiPISA2

OGRENCIID OKUMA_ZEVK ODOKUMA1
1 79200768  -0.2891    376.022 
2 79201064   0.6040    512.316 
3 79201275  -1.1538    393.006 
4 79201481   0.6669    552.457 
5 79201556   0.3568    441.286 
6 79201652  -0.0886    410.648

Elimizde bulunan midiPISA1 ve midiPISA2 veri setlerini kullanarak left_join() fonksiyonu ile birleştirme işlemi uygulanır.

In [58]:
midiPISA1%>% left_join(midiPISA2) #veri birleştirme

Joining with `by = join_by(OGRENCIID)`


OGRENCIID CINSIYET ST097Q01TA ST097Q02TA OKUMA_ZEVK ODOKUMA1
1 79200768  2        1          2          -0.2891    376.022 
2 79201064  2        3          2           0.6040    512.316 
3 79201118  1        2          3               NA         NA 
4 79201275  2        2          2          -1.1538    393.006 
5 79201481  2        3          3           0.6669    552.457 
6 79201556  2        3          3           0.3568    441.286

Elde edilen çıktı incelendiğinde, midiPISA1 verisindeki tüm satırlar, midiPISA2 verisi ile eşleştirilmiş, “79201652”ıd numaralı öğrenci midiPISA1 de olmadığından hiç alınmamıştır. “79201118” ıd numaralı öğrenci ise midiPISA2 verisinde yer almadığında okumaktan zevk alma ve okuma olası değer1 değişkenleri “NA” olarak oluşturulmuştur.

#**right_join()**

A %>% left_join(B) ile B verisindeki tüm satırlar, mümkün olduğunda A verisi ile eşleştirilerek (olmadığında “NA” verir), hem A hem de B den gelen sütunlar alınır.

right_join() fonksiyonu ile öncelikle X2 değişkenine ait tüm gözlemler alınmıştır. X1 değişkenine ait ID numarası “3” olan gözlem olmadığı için yeni veri setinde bu kısım kayıp veri(NA) olaraK girilmiştir. Burada da left_join() fonksiyonunda olduğu gibi birleştirme yapılacak iki veri setine ait ortak bir değişkenin (“ID”) olması önemlidir.

Örnek veri seti üzerinden açıklayalım:

In [59]:
midiPISA1%>% right_join(midiPISA2) # veri birleştirme

Joining with `by = join_by(OGRENCIID)`


OGRENCIID CINSIYET ST097Q01TA ST097Q02TA OKUMA_ZEVK ODOKUMA1
1 79200768   2        1          2         -0.2891    376.022 
2 79201064   2        3          2          0.6040    512.316 
3 79201275   2        2          2         -1.1538    393.006 
4 79201481   2        3          3          0.6669    552.457 
5 79201556   2        3          3          0.3568    441.286 
6 79201652  NA       NA         NA         -0.0886    410.648

lde edilen çıktı incelendiğinde, midiPISA2 verisindeki tüm satırlar, midiPISA1 verisi ile eşleştirilmiş, “79201118”ıd numaralı öğrenci midiPISA2 de olmadığından oluşturulan veri setine hiç alınmamıştır.”79201652” ıd numaralı öğrenci ise midiPISA1 verisinde yer almadığında CINSIYET, ST097Q01TA,ST097Q01TA değişkenleri “NA” olarak oluşturulmuştur. Görüldüğü üzere left_join() ile right_join() fonksiyonları arasındaki fark eşleştirme öncesi temel alınacak ana veri setinin farklı olmasıdır. Fakat pratikte genellikle left_join() fonksiyonu kullanılır.

# **inner_join()**

A %>% inner_join(B) ile sadece A ve B nin eşleşen satırlarını birleştirir. Yani hem A hem de B den gelen sütunları alır.

inner_join() fonksiyonu ile öncelikle “ID” değişkeni baz alınarak ortak olan satırlar yani 2. satır birleştirilir.

In [60]:
midiPISA1%>% inner_join(midiPISA2) # veri birleştirme

Joining with `by = join_by(OGRENCIID)`


OGRENCIID CINSIYET ST097Q01TA ST097Q02TA OKUMA_ZEVK ODOKUMA1
1 79200768  2        1          2          -0.2891    376.022 
2 79201064  2        3          2           0.6040    512.316 
3 79201275  2        2          2          -1.1538    393.006 
4 79201481  2        3          3           0.6669    552.457 
5 79201556  2        3          3           0.3568    441.286

“OGRENCIID” değişkeni ortak değişken olduğundan bu değişkene göre birleştirme işlemi yapılmış ve sadece her iki veri setinde yer alan öğrenciler seçilmiştir. Bu durumda “79201118” ve “79201652” id numaraya sahip öğrenciler birleştirme işleminden sonra yeni veri setinde yer almamıştır. Böylece yeni veri seti beş satırdan oluşmuştur.

# **full_join()**

A %>% full_join(B) ile A ve B veri setinde yer alan tüm satırları birleştirir. Hem A hem de B den gelen sütunları alır.


In [61]:
midiPISA1%>% full_join(midiPISA2) # veri birleştirme

Joining with `by = join_by(OGRENCIID)`


OGRENCIID CINSIYET ST097Q01TA ST097Q02TA OKUMA_ZEVK ODOKUMA1
1 79200768   2        1          2         -0.2891    376.022 
2 79201064   2        3          2          0.6040    512.316 
3 79201118   1        2          3              NA         NA 
4 79201275   2        2          2         -1.1538    393.006 
5 79201481   2        3          3          0.6669    552.457 
6 79201556   2        3          3          0.3568    441.286 
7 79201652  NA       NA         NA         -0.0886    410.648

#**semi_join()**

A %>% semi_join(B) ile A veri setinin B ile eşleşen satırları alınarak sadece A dan gelen sütunlar yeni veri setinde yer alır.

semi_join() fonksiyonu ile öncelikle “ID” değişkeni baz alınarak ortak olan satırlar yani 2. satır birleştirilir. Fakat yine sadece A veri setinde yer alan değişken ya da değişkenler alınır.

In [62]:
midiPISA1%>% semi_join(midiPISA2)

Joining with `by = join_by(OGRENCIID)`


OGRENCIID CINSIYET ST097Q01TA ST097Q02TA
1 79200768  2        1          2         
2 79201064  2        3          2         
3 79201275  2        2          2         
4 79201481  2        3          3         
5 79201556  2        3          3

Çıktıda midiPISA1 verisinin midiPISA2 verisi ile eşleşen satırları ele alındığından “79201118” id numaralı öğrenci yeni veri setine dahil edilmemiştir. Ayrıca yeni veri setinde sadece midiPISA1 değişkeninde yer alan CINSIYET, ST097Q01TA, ST097Q02TA değişkenleri yer almıştır.

# **anti_join()**

A %>% anti_join(B) ile A’nın B ile eşleşemeyen satırları alınarak yeni veri setinde sadece A’dan gelen sütunlara yer verilir.

anti_join() fonksiyonu ile öncelikle “ID” değişkeni baz alınarak ortak olmayan satırlar yani sadece 1. satır ve A veri setindeki değişken (X1) alınmıştır.

In [63]:
midiPISA1%>% anti_join(midiPISA2) # veri birleştirme

Joining with `by = join_by(OGRENCIID)`


OGRENCIID CINSIYET ST097Q01TA ST097Q02TA
1 79201118  1        2          3

Çıktıda midiPISA1 verisinin midiPISA2 verisi ile eşleşmeyen satırları ele alındığından “79201118” id numaralı öğrenci yeni veri setine dahil edilmiştir. Ayrıca yeni veri setinde sadece midiPISA1 değişkeninde yer alan CINSIYET, ST097Q01TA, ST097Q02TA değişkenleri yer almıştır.

Not: Aynı değerleri içeren satırların olduğu sütunların eşleşmesi gerektiğini söyledik. Bunları birleştirme için aslında bir by = argümanı da kullanılır. Fakat birleştirme yapmak istediğimiz sütun/ların isimleri aynı olduğunda by = argümanını kullanmaya gerek kalmaz. Buraya kadar yapılan örnekler aşağıdaki gibi pipe operatörü kullanIlmadan da yapılabilir.

In [64]:
left_join(midiPISA1,midiPISA2, by=join_by(OGRENCIID))

OGRENCIID CINSIYET ST097Q01TA ST097Q02TA OKUMA_ZEVK ODOKUMA1
1 79200768  2        1          2          -0.2891    376.022 
2 79201064  2        3          2           0.6040    512.316 
3 79201118  1        2          3               NA         NA 
4 79201275  2        2          2          -1.1538    393.006 
5 79201481  2        3          3           0.6669    552.457 
6 79201556  2        3          3           0.3568    441.286

Elde edilen çıktı midiPISA1 %>% left_join(midiPISA2) fonksiyonu ile elde edilen çıktı ile aynıdır. Fakat değişken isimleri farklı olduğunda by argümanını kullanmak anlamlı olabilir. Şimdi bununla ilgili bir örnek yapalım. Öncelikle midiPISA2 veri setindeki “OGRENCIID” değişken ismini “STDNTID” olarak değiştirelim.

In [65]:
midiPISA3<-midiPISA2 %>%
  rename(STDNTID=OGRENCIID)
head(midiPISA3)

STDNTID  OKUMA_ZEVK ODOKUMA1
1 79200768 -0.2891    376.022 
2 79201064  0.6040    512.316 
3 79201275 -1.1538    393.006 
4 79201481  0.6669    552.457 
5 79201556  0.3568    441.286 
6 79201652 -0.0886    410.648

midiPISA3 verisi aslında midiPISA2 verisinin aynısıdır sadece tek bir değişkenin ismi değişmiştir.

Veri setinde eşleştirme yapılması istenilen değişkenin farklı adları olduğunda aşağıdaki kod kullanılabilir.

In [66]:
left_join(midiPISA2, midiPISA3, by=c("OGRENCIID"="STDNTID"))

OGRENCIID OKUMA_ZEVK.x ODOKUMA1.x OKUMA_ZEVK.y ODOKUMA1.y
1 79200768  -0.2891      376.022    -0.2891      376.022   
2 79201064   0.6040      512.316     0.6040      512.316   
3 79201275  -1.1538      393.006    -1.1538      393.006   
4 79201481   0.6669      552.457     0.6669      552.457   
5 79201556   0.3568      441.286     0.3568      441.286   
6 79201652  -0.0886      410.648    -0.0886      410.648

Bu bölümde farklı veri setlerini birleştirme işlemlerine yer verilmiştir. Veri Düzenleme IV bölümünde veri setlerini farklı formatlara getirme ve analize hazırlık işlemlerine değinilecektir.

Bir veriyi R ortamına aktardıktan sonra veri setinde yer alan tüm satır ya da sütunların doğru bir şekilde aktarılıp aktarılmadığı, değişken isimlerinin düzgün olup olmadığı yani özellikle sütun adlarında boşluk olmaması ya da farklı karakterler bulunmaması kontrol edilmelidir. İlk olarak R ortamına aktarılan boş satır ve sütunlar olup olmadığı filter() ve select() gibi fonksiyonlarla incelenebilir. Eksik verilerin nasıl temsil edildiği kontrol edilmelidir. NA,” “ (bosluk), ., 999 , 9999 vb. şekilde ifade edilen eksik veriler mutate() ve ifelse() ile düzenlenebilir. Ayrıca karakter (character) ve faktör (factor) değişkenlerinin de düzgün tanımlanıp tanımlanmadığı incelenmelidir.

#**gather()**

gather() fonksiyonu bir dizi sütun alır ve onları iki yeni sütuna (kendi adını verebileceğin) dönüştürür.

Fonksiyonun kullanım şekli;

gather(data, key, value, ..., na.rm = FALSE,
       convert = FALSE, factor_key = FALSE)

A Key: Orijinal sütun adlarını saklayan bir anahtar.

A value: Bu orijinal sütunlardaki değerlere sahip bir değer.
inde çalışalım.

In [67]:
genisveri <- midiPISA %>% select(OGRENCIID,ODOKUMA1:ODOKUMA5) #belli değişkenlerin seçilmesi
genisveri %>% head(6) # verinin ilk 6 satırının görüntülenmesi

OGRENCIID ODOKUMA1 ODOKUMA2 ODOKUMA3 ODOKUMA4 ODOKUMA5
1 79200768  376.022  417.746  420.630  413.837  434.434 
2 79201064  512.316  473.229  563.902  485.396  500.394 
3 79201118  396.383  413.859  423.121  452.124  392.434 
4 79201275  393.006  428.757  364.850  382.522  378.563 
5 79201481  552.457  570.109  562.955  530.835  532.487 
6 79201556  441.286  415.682  406.586  437.001  473.036

Elde edilen çıktıda öğrenci ıd ve beş okuma olası değerinin yer aldığı toplam altı değişkenden oluşan veri seti görüntülenmektedir. Bu değişkenler sütunlarda yer almaktadır. gather() fonksiyonu geniş veriyi, uzun veri haline getirir.

In [68]:
#gather() fonksiyonu eski kaldığı için onun yerine pivot_longer fonksiyonunu kullandık
library(tidyr)
library(dplyr)

uzun <- genisveri %>%
  pivot_longer(
    cols = ODOKUMA1:ODOKUMA5,
    names_to = "O_OD",
    values_to = "okumapuan"
  )

uzun %>% arrange(OGRENCIID) %>% head(10)

OGRENCIID O_OD     okumapuan
1  79200001  ODOKUMA1 449.876  
2  79200001  ODOKUMA2 457.572  
3  79200001  ODOKUMA3 413.428  
4  79200001  ODOKUMA4 430.044  
5  79200001  ODOKUMA5 439.072  
6  79200002  ODOKUMA1 669.156  
7  79200002  ODOKUMA2 665.848  
8  79200002  ODOKUMA3 684.978  
9  79200002  ODOKUMA4 664.624  
10 79200002  ODOKUMA5 659.670

Elde edilen çıktıda öğrenci ıd ve beş okuma olası değerinin yer aldığı toplam altı değişkenden oluşan veri seti görüntülenmektedir. Bu değişkenler sütunlarda yer almaktadır. gather() fonksiyonu geniş veriyi, uzun veri haline getirir.

Çıktı incelendiğinde, oluşan veride ODOKUMA1, ODOKUMA2, ODOKUMA3, ODOKUMA4 ve ODOKUMA5 okumapuanı değişkeninin değerleri hâline gelmiştir. Çıktıda görüldüğü gibi, şimdi ID dışında iki sütunumuz var: Biri kategorik diğeri sayısal değerleri içerir. Her katılımcı icin beş farklı okuma olası değeri olduğu için her bir ID değeri beş kere tekrarlanmaktadır. Burada veri setinin ilk on satırı görüntülendiğinden sadece 79200001 ve 79200002 id numaralı öğrenciler görüntülenmektedir.

#**spread()**

spread() fonksiyonu uzun veriden tekrar geniş veri olusturmaya yarar. gather() fonksiyonunun tersi olan işlevi yapar

In [69]:
tekrar_genis <- uzun %>% spread(O_OD,okumapuan) # geniş veri oluşturulması
tekrar_genis %>% head(6) # ilk altı satırın görüntülenmesi

OGRENCIID ODOKUMA1 ODOKUMA2 ODOKUMA3 ODOKUMA4 ODOKUMA5
1 79200001  449.876  457.572  413.428  430.044  439.072 
2 79200002  669.156  665.848  684.978  664.624  659.670 
3 79200003  451.875  502.208  443.832  456.082  436.965 
4 79200004  346.944  317.129  339.054  325.048  367.116 
5 79200005  466.959  497.879  415.394  470.793  446.268 
6 79200006  366.303  364.383  384.399  420.318  351.426

#**pivot_longer() ve pivot_wider()**

Verilerin girilme şekli genellikle geniş ve uzun olmak üzere iki formattan oluşur. Geniş formatta veriler, bir gözlemin özellikleri veya yanıtlar tek bir satırda verilir. Genellikle veriler bu şekilde girilmesine rağmen geniş format her zaman kullanışlı olmayabilir. Geniş verinin uzun veriye dönüştürülmesini gather() ve uzun verinin geniş veriye dönüşütürülmesini spread()fonksiyonu ile gerçekleştirdik. Ancak bahsedilen iki fonksiyona alternatif yeni fonksiyonlar üretilmiştir. Bu bölümde bu iki fonksiyon açıklanacaktır.

midiPISA verisetinden daha az değişken içerecek şekilde bir geniş veri seti örneği oluşturalım.

In [70]:
genisveri <- midiPISA %>% select(OGRENCIID,ODOKUMA1:ODOKUMA5) #belli değişkenlerin seçilmesi
genisveri %>% head(6) # verinin ilk 6 satırının görüntülenmesi

OGRENCIID ODOKUMA1 ODOKUMA2 ODOKUMA3 ODOKUMA4 ODOKUMA5
1 79200768  376.022  417.746  420.630  413.837  434.434 
2 79201064  512.316  473.229  563.902  485.396  500.394 
3 79201118  396.383  413.859  423.121  452.124  392.434 
4 79201275  393.006  428.757  364.850  382.522  378.563 
5 79201481  552.457  570.109  562.955  530.835  532.487 
6 79201556  441.286  415.682  406.586  437.001  473.036

Elde edilen çıktıda öğrenci ıd ve beş okuma olası değerinin yer aldığı toplam altı değişkenden yer alan veri seti görüntülenmektedir. Bu değişkenler sütunlarda yer almakta olup geniş veri formatındadır. pivot_longer fonksiyonu geniş veriyi, uzun veri haline getirir.

In [71]:
uzun <- genisveri %>% pivot_longer(names_to="okumapuan",values_to="deger",cols=ODOKUMA1:ODOKUMA5)
uzun %>% head(5)

OGRENCIID okumapuan deger  
1 79200768  ODOKUMA1  376.022
2 79200768  ODOKUMA2  417.746
3 79200768  ODOKUMA3  420.630
4 79200768  ODOKUMA4  413.837
5 79200768  ODOKUMA5  434.434

Çıktı incelendiğinde, oluşan veride ODOKUMA1, ODOKUMA2, ODOKUMA3, ODOKUMA4 ve ODOKUMA5 ayrı bir sütunun değerleri haline gelmiştir. Bu okuma puan türlerinin sütununun yer aldığı değişken names_to argümanı ile “okumapuan” olarak isimlendirilmiştir. Ayrıca values_to argümanı ise okuma puanı değerlerinin yer aldığı sütun isimlendirilmiştir. Çıktıda görüldüğü gibi, şimdi ID dışında iki sütunumuz var: Biri okuma puanı türü için, diğeri okuma puanı türleri için. Her katılımcı icin beş farklı okuma olası değeri olduğu için her bir ID değeri beş kere tekrarlanmaktadır. Burada veri setinin ilk beş satırı görüntülendiğinden sadece 792200768 id numaralı öğrencinin değerleri görüntülenmektedir.

Bir veri setini daha iyi yorumlayabilmek amacıyla uzun veri formatından geniş veri formatına dönüştürülür. Genellikle bir gözlem için değerlerin birden çok satırda yer aldığı durumlarda tercih edilir. Bunun için pivot_wider() fonksiyonu kullanılır.

In [72]:
genis<- uzun %>% pivot_wider(names_from="okumapuan",values_from="deger")
genis %>% head(5)

OGRENCIID ODOKUMA1 ODOKUMA2 ODOKUMA3 ODOKUMA4 ODOKUMA5
1 79200768  376.022  417.746  420.630  413.837  434.434 
2 79201064  512.316  473.229  563.902  485.396  500.394 
3 79201118  396.383  413.859  423.121  452.124  392.434 
4 79201275  393.006  428.757  364.850  382.522  378.563 
5 79201481  552.457  570.109  562.955  530.835  532.487

#**separate()**

separate() fonksiyonu bir sütunu birden çok sütuna ayırır. Değerlerin sütun adlarına gömüldüğü toplanmış verilerde ortaktır. Oluşan veride okuma puanı değerlerinin karakter ve sayısal değerlerini ayırmak için separate()fonkisyonu kullanılabilir.

In [73]:
uzun_v1 <- uzun %>% separate(okumapuan, c("OD","Sayi"),"MA") # bir sütunu iki sütuna ayırma

uzun_v1 %>% head(3) #ilk üç satırın görüntülenmesi

OGRENCIID OD    Sayi deger  
1 79200768  ODOKU 1    376.022
2 79200768  ODOKU 2    417.746
3 79200768  ODOKU 3    420.630

Elde edilen çıktıya göre, okuma puanı olası değerlerinin yer aldığı sütun ikiye ayrılarak OD sütunu ve sayı sütunundan oluşmaktadır. Çıktının ilk üç satırı görüntülendiği için tek bir öğrenciye ait üç olası değerler yer almaktadır.

#**unite()**
separate() fonksiyonunun tam tersi olarak iki sütunu alıp tek sütunda birleştirir.

In [74]:
uzun_birles <- uzun_v1 %>% unite(ODOKUMA, OD, Sayi, sep = "_") # sütun birleştirmenin yapılması
uzun_birles %>% head(3) # ilk üç satırın görüntülenmesi

OGRENCIID ODOKUMA deger  
1 79200768  ODOKU_1 376.022
2 79200768  ODOKU_2 417.746
3 79200768  ODOKU_3 420.630

Elde edilen çıktı incelendiğinde, öğrenci id değişkeni hariç iki sütunun olduğu görülmektedir. ODOKUMA sütunu, okuma puanlarının isimlerinden, değer ise okuma olası puanı değerlerinden oluşmaktadır.

separate() fonksiyonunun alternatifi extract() ve unite() fonksiyonları ile yapılabilecek olan işlemler mutate() fonksiyonu ile de yapılabilir.

Bu alternatiflerin uygunluğunun özel kullanım durumunuza ve verilerinizin niteliğine bağlı olduğunu unutmayın. Paketler zaman içinde yeni fonksiyonlara veya iyileştirmelere sahip olabileceğinden, en son güncellemeler için her zaman fonksiyon yardım sayfalarını kontrol etmenizi öneriyoruz.